internal/topology/variables/clusterclass_variable_validation.go

```diff
@@ -215,9 +233,172 @@ func validateSchema(schema *apiextensions.JSONSchemaProps, fldPath *field.Path)
+		}
+		if hasNewlines(rule.FieldPath) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("x-kubernetes-validations").Index(i).Child("fieldPath"), rule.FieldPath, "fieldPath must not contain line breaks"))
+		}
+		if rule.FieldPath != "" {
+			if !pathValid(schema, rule.FieldPath) {
+				allErrs = append(allErrs, field.Invalid(fldPath.Child("x-kubernetes-validations").Index(i).Child("fieldPath"), rule.FieldPath, "fieldPath must be a valid path"))
+			}
+		}
+	}
+
+	// If any schema related validation errors have been found at this level or deeper, skip CEL expression validation.
+	// Invalid OpenAPISchemas are not always possible to convert into valid CEL DeclTypes, and can lead to CEL
+	// validation error messages that are not actionable (will go away once the schema errors are resolved) and that
+	// are difficult for CEL expression authors to understand.
+	if len(allErrs) > 0 {
+		return allErrs
+	}
+
+	celContext := apiextensionsvalidation.RootCELContext(schema)
+	allErrs = append(allErrs, validateCELExpressions(schema, fldPath, celContext)...)
+
+	return allErrs
+}
+
+var newlineMatcher = regexp.MustCompile(`[\n\r]+`) // valid newline chars in CEL grammar
+func hasNewlines(s string) bool {
+	return newlineMatcher.MatchString(s)
+}
+
+func pathValid(schema *apiextensions.JSONSchemaProps, path string) bool {
+	// To avoid duplicated code and better maintain, using ValidFieldPath func to check if the path is valid
+	if ss, err := structuralschema.NewStructural(schema); err == nil {
+		_, _, err := cel.ValidFieldPath(path, ss)
+		return err == nil
+	}
+	return true
+}
+
+func validateCELExpressions(schema *apiextensions.JSONSchemaProps, fldPath *field.Path, celContext *apiextensionsvalidation.CELSchemaContext) field.ErrorList {
+	var allErrs field.ErrorList
+
+	if schema.AdditionalProperties != nil {
+		allErrs = append(allErrs, validateCELExpressions(schema.AdditionalProperties.Schema, fldPath.Child("additionalProperties"), celContext.ChildAdditionalPropertiesContext(schema.AdditionalProperties.Schema))...)
+	}
+
+	if len(schema.Properties) != 0 {
+		for property, propertySchema := range schema.Properties {
+			p := propertySchema
+			allErrs = append(allErrs, validateCELExpressions(&p, fldPath.Child("properties").Key(property), celContext.ChildPropertyContext(&p, property))...)
+		}
+	}
+
 	if schema.Items != nil {
 		allErrs = append(allErrs, validateSchema(schema.Items.Schema, fldPath.Child("items"))...)
+		allErrs = append(allErrs, validateCELExpressions(schema.Items.Schema, fldPath.Child("items"), celContext.ChildItemsContext(schema.Items.Schema))...)
+	}
+
+	typeInfo, err := celContext.TypeInfo()
+	if err != nil {
+		return append(allErrs, field.InternalError(fldPath.Child("x-kubernetes-validations"), errors.Wrap(err, "internal error: failed to construct type information for x-kubernetes-validations rules")))
+	}
+	if typeInfo == nil {
```

Comment:
This seems to be an error case in upstream
> allErrs.CELErrors = append(allErrs.CELErrors, field.InternalError(fldPath.Child("x-kubernetes-validations"), fmt.Errorf("internal error: failed to retrieve type information for x-kubernetes-validations")))

Reply:
I've reverted that change now and now ignore if typeInfo is nil. This passes all existing tests, but honestly I don't know why this would be nil for CAPI variables and non-nil upstream. Any ideas?

Reply:
I'll take another look

Reply:
Okay I know why. They only run any of this if `len(schema.XValidations) > 0`. In this case it is indeed an error if typeInfo is nil (typeInfo is not expected to be nil if there are XValidations)